setup/mscs/msc2946/msc2946.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2021 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package msc2946 'Spaces Summary' implements https://github.com/matrix-org/matrix-doc/pull/2946
+package msc2946
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/mux"
+	chttputil "github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/internal/hooks"
+	"github.com/matrix-org/dendrite/internal/httputil"
+	roomserver "github.com/matrix-org/dendrite/roomserver/api"
+	"github.com/matrix-org/dendrite/setup"
+	userapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	ConstCreateEventContentKey = "org.matrix.msc1772.type"
+	ConstSpaceChildEventType   = "org.matrix.msc1772.space.child"
+	ConstSpaceParentEventType  = "org.matrix.msc1772.room.parent"
+)
+
+// SpacesRequest is the request body to POST /_matrix/client/r0/rooms/{roomID}/spaces
+type SpacesRequest struct {
+	MaxRoomsPerSpace int    `json:"max_rooms_per_space"`
+	Limit            int    `json:"limit"`
+	Batch            string `json:"batch"`
+}
+
+// Defaults sets the request defaults
+func (r *SpacesRequest) Defaults() {
+	r.Limit = 100
+	r.MaxRoomsPerSpace = -1
+}
+
+// SpacesResponse is the response body to POST /_matrix/client/r0/rooms/{roomID}/spaces
+type SpacesResponse struct {
+	NextBatch string `json:"next_batch"`
+	// Rooms are nodes on the space graph.
+	Rooms []Room `json:"rooms"`
+	// Events are edges on the space graph, exclusively m.space.child or m.room.parent events
+	Events []gomatrixserverlib.ClientEvent `json:"events"`
+}
+
+// Room is a node on the space graph
+type Room struct {
+	gomatrixserverlib.PublicRoom
+	NumRefs  int    `json:"num_refs"`
+	RoomType string `json:"room_type"`
+}
+
+// Enable this MSC
+func Enable(
+	base *setup.BaseDendrite, rsAPI roomserver.RoomserverInternalAPI, userAPI userapi.UserInternalAPI,
+) error {
+	db, err := NewDatabase(&base.Cfg.MSCs.Database)
+	if err != nil {
+		return fmt.Errorf("Cannot enable MSC2946: %w", err)
+	}
+	hooks.Enable()
+	hooks.Attach(hooks.KindNewEventPersisted, func(headeredEvent interface{}) {
+		he := headeredEvent.(*gomatrixserverlib.HeaderedEvent)
+		hookErr := db.StoreReference(context.Background(), he)
```

Comment:
I realise now that filtering takes place in `StoreReference`, but that wasn't immediately obvious and left me wondering why we'd try to find references in all timeline events. Might make sense to perform the filtering here instead, or at the very least comment what's going on, as leaving the storage API to filter what we give it isn't really a pattern we use elsewhere.

Reply:
Will tweak this in the next PR.